tests/integration/cocoon/test_cocoon_transactions.py

```diff
@@ -819,3 +820,92 @@ def test_filter_unmatched_transactions_can_paginate_responses_for_2001_transacti
+                    "transaction_id": "id",
+                    "type": "transaction_type",
+                    "transaction_date": "transaction_date",
+                    "settlement_date": "settlement_date",
+                    "units": "units",
+                    "transaction_price.price": "transaction_price",
+                    "transaction_price.type": "price_type",
+                    "total_consideration.amount": "amount",
+                    "total_consideration.currency": "trade_currency",
+                },
+                {"transaction_currency": "trade_currency"},
+                {
+                    "Isin": "isin",
+                    "Figi": "figi",
+                    "ClientInternal": "client_internal",
+                    "Currency": "currency_transaction",
+                },
+                ["exposure_counterparty", "compls", "val", "location_region"],
+                "load_dataframe_test",
+                None,
+                ["SHK_data"],
+                ['Transaction/load_dataframe_test/SHK_data']
+            ]
+        ]
+    )
+    def test_load_from_dataframe_non_existent_subholding_keys(
+        self,
+        _,
+        scope,
+        file_name,
+        mapping_required,
+        mapping_optional,
+        identifier_mapping,
+        property_columns,
+        properties_scope,
+        batch_size,
+        sub_holding_keys,
+        expected_sub_holdings_keys
+    ):
+        """
+        This checks whether load_from_data_frame creates subholding keys for transactions when they don't already exist.
+        """
+        portfolios = self.api_factory.build(lusid.PortfoliosApi).list_portfolios_for_scope(scope).values
+
+        for portfolio in portfolios:
+            self.api_factory.build(lusid.PortfoliosApi).delete_portfolio(scope, portfolio.id.code)
+
+        self.api_factory.build(lusid.TransactionPortfoliosApi).create_portfolio(scope, {'displayName': "test_load_from_dataframe_non_existent_subholding_keys portfolio", 'code': 'no-SHK', 'baseCurrency': 'GBP', 'created': "2017-06-22T00:00:00.0000000+00:00"})
+
+        data_frame = pd.read_csv(Path(__file__).parent.joinpath(file_name))
+
+        cocoon.cocoon.load_from_data_frame(
+            api_factory=self.api_factory,
+            scope=scope,
+            data_frame=data_frame,
+            mapping_required=mapping_required,
+            mapping_optional=mapping_optional,
+            file_type="transactions",
+            identifier_mapping=identifier_mapping,
+            property_columns=property_columns,
+            properties_scope=properties_scope,
+            batch_size=batch_size,
+            sub_holding_keys = sub_holding_keys
+        )
+
+        portfolio_details = self.api_factory.build(
+            lusid.api.TransactionPortfoliosApi
+        ).get_details(scope=scope, code='no-SHK')
+
+        transactions = self.api_factory.build(
+            lusid.api.TransactionPortfoliosApi
```

Comment:
A repeated build of the API, just build it once, put to a variable and call the functions on that variable